### ISSUE TITLE
M2 MacBook keyboard, trackpad do not work when booting from USB

### ISSUE THREAD

**Issue by starThorn** [AsahiLinux/asahi-installer#213]
I have a boot stub and nixos rootfs on the USB. The bootstub chainloads to an EFI partition on the M2's internal NVMe. Booting works fine; however, from Uboot on the M2's trackpad and keyboard stop working and I have to use an external keyboard and mouse. Any ideas why?

With the understanding that expert mode is for experimental use, I would also like to point out that running the installer in expert mode when installing onto an external drive fails out here: https://github.com/AsahiLinux/asahi-installer/blob/e4d8bb3524fd004abac7468aac275f33d5c94c56/src/osinstall.py#L123

**Comment by angelbirth** [comment]
I'm sorry for asking, have you compile the kernel with the correct config (i.e. drivers for keyboard and trackpad)?

**Comment by starThorn** [comment]
@angelbirth valid concern, but I do not think that is the issue. I have an additional boot stub located on the internal NVMe, that chainloads to the same EFI partition, which then uses the partition on my external drive containing the NixOS rootfs. So when I boot from this stub, it uses the same EFI and same rootfs as when I boot from the boot stub on the external drive. The strange thing with this internal boot stub is that the keyboard and trackpad work from Uboot on, including with NixOS. 

In summary, the keyboard and trackpad work for the same EFI and  rootfs if you use a boot stub on the internal NVMe.

**Comment by jannau** [comment]
which macOS version(s) are used for the boot stubs? which m1n1, u-boot and kernel versions are in play? for u-boot, which exact git sha1? If u-boot is not asahi-v2023.07.02-2 and one of the stubs is macOS 13.5 based keyboard and trackpad will not work.

**Comment by starThorn** [comment]
@jannau From what you pointed out, it is likely that I am using an outdated EFI. When using the Asahi Installer (expert mode) to put the boot stub on the drive, I chose the 13.5 macOS version. On the other hand, the boot stub installed with the regular installer put a 12.4 boot stub and EFI partitions on my internal NVMe. So it is likely that when I combine the 13.5 boot stub with the 12.4 EFI, things are not working.

So, how would you recommend correcting the EFI to be up to date? From the [documentation ](https://github.com/AsahiLinux/docs/wiki/Distro:Boot-process-guide#build-instructions), it would seem that I need to build the default branches of m1n1 and u-boot, and then concatenate them together. I assume I can leave out the device tree since I plan to use NixOS?  Then I figure I would replace the boot.bin on my mounted EFI with this custom m1n1.bin -- is this process correct or am I missing something?

**Comment by starThorn** [comment]
Resolution: the issue was with https://github.com/AsahiLinux/PKGBUILDs/blob/main/uboot-asahi/PKGBUILD#L4 being set for an old version. Once I built U-boot locally following the documentation referenced above, and re-concatenated m1n1 with the dtbs and uboot, I was able to use the keyboard and trackpad successfully.
